File: backend/app/utils/nutrition.py
# backend/app/utils/nutrition.py

from typing import Tuple, Dict
from app.models import User

# Caloric multipliers for activity levels
TDEE_MULTIPLIERS = {
    "low":      1.2,
    "moderate": 1.375,
    "high":     1.55,
}

# ~1100 kcal/week per kg → ≈157 kcal/day per kg
CAL_ADJUST_PER_KG = 1100 / 7  

# Macro grams per kg of bodyweight by goal
MACROS_G_PER_KG = {
    "gain muscle": {"protein": 2.0, "carbs": 4.0, "fat": 1.0},
    "maintenance": {"protein": 1.8, "carbs": 3.0, "fat": 0.8},
    "fat loss":    {"protein": 2.2, "carbs": 2.5, "fat": 0.8},
    "weight loss": {"protein": 2.0, "carbs": 2.0, "fat": 0.8},
}

def compute_bmr(sex: str, age: int, weight: float, height: float,
                height_unit: str, weight_unit: str) -> float:
    # Mifflin-St Jeor
    w_kg = weight if weight_unit == "kg" else weight * 0.453592
    h_cm = height if height_unit == "cm" else height * 2.54
    if sex.lower() == "male":
        return 10 * w_kg + 6.25 * h_cm - 5 * age + 5
    else:
        return 10 * w_kg + 6.25 * h_cm - 5 * age - 161

def compute_nutrition_profile(user: User) -> Tuple[int, Dict[str,int]]:
    # BMR → TDEE
    bmr = compute_bmr(user.sex, user.age, user.weight,
                      user.height, user.height_unit, user.weight_unit)
    tdee = bmr * TDEE_MULTIPLIERS.get(user.activity_level or "moderate", 1.375)

    # Caloric adjustment toward target
    if user.weight_target is not None:
        cur_w = user.weight if user.weight_unit == "kg" else user.weight * 0.453592
        tgt_w = user.weight_target if user.weight_target_unit == "kg" else user.weight_target * 0.453592
        delta_kg = tgt_w - cur_w
        cal_adjust = CAL_ADJUST_PER_KG * delta_kg
    else:
        cal_adjust = 0

    maintenance_cals = int(round(tdee + cal_adjust))

    # 3) Macro targets
    goal_key = user.goal.lower() if user.goal else "maintenance"
    grams_per_kg = MACROS_G_PER_KG.get(goal_key, MACROS_G_PER_KG["maintenance"])
    w_kg = user.weight if user.weight_unit == "kg" else user.weight * 0.453592

    macro_targets = {
        k: int(round(grams_per_kg[k] * w_kg))
        for k in grams_per_kg
    }

    return maintenance_cals, macro_targets

File: frontend/src/utils/nutrition.ts
// src/utils/nutrition.ts

export const TDEE_MULTIPLIERS: Record<string, number> = {
    low:      1.2,
    moderate: 1.375,
    high:     1.55,
  };
  
  export const MACROS_G_PER_KG: Record<string, Record<string, number>> = {
    'gain muscle':   { protein: 2.0, carbs: 4.0, fat: 1.0 },
    'maintenance':   { protein: 1.8, carbs: 3.0, fat: 0.8 },
    'fat loss':      { protein: 2.2, carbs: 2.5, fat: 0.8 },
    'weight loss':   { protein: 2.0, carbs: 2.0, fat: 0.8 },
  };
  
  export function computeBMR(
    sex: string,
    age: number,
    weight: number,
    height: number,
    height_unit: string,
    weight_unit: string
  ): number {
    const w_kg = weight_unit === 'kg' ? weight : weight * 0.453592;
    const h_cm = height_unit === 'cm' ? height : height * 2.54;
    if (sex.toLowerCase() === 'male') {
      return 10 * w_kg + 6.25 * h_cm - 5 * age + 5;
    } else {
      return 10 * w_kg + 6.25 * h_cm - 5 * age - 161;
    }
  }
  
  export function computeNutritionProfile({
    sex,
    age,
    weight,
    height,
    height_unit,
    weight_unit,
    activity_level,
    goal,
  }: {
    sex: string;
    age: number;
    weight: number;
    height: number;
    height_unit: string;
    weight_unit: string;
    activity_level: string;
    goal?: string | null;
  }): {
    maintenanceCalories: number;
    macroTargets: { protein: number; carbs: number; fat: number };
  } {
    // 1) BMR → TDEE
    const bmr = computeBMR(sex, age, weight, height, height_unit, weight_unit);
    const tdee = bmr * TDEE_MULTIPLIERS[activity_level || 'moderate'];
  
    // 2) Goal adjustment
    const g = (goal || 'maintenance').toLowerCase();
    let goalCals: number;
    if (g === 'fat loss') {
      goalCals = tdee - 200;
    } else if (g === 'weight loss') {
      goalCals = tdee - 500;
    } else if (g === 'gain muscle') {
      goalCals = tdee + 200;
    } else {
      goalCals = tdee;
    }
    const maintenanceCalories = Math.round(goalCals);
  
    // 3) Macro targets
    const gramsPerKg = MACROS_G_PER_KG[g] ?? MACROS_G_PER_KG['maintenance'];
    const w_kg = weight_unit === 'kg' ? weight : weight * 0.453592;
    const macroTargets = {
      protein: Math.round(gramsPerKg.protein * w_kg),
      carbs:   Math.round(gramsPerKg.carbs * w_kg),
      fat:     Math.round(gramsPerKg.fat * w_kg),
    };
  
    return { maintenanceCalories, macroTargets };
  }  